_studio/mfx_lib/shared/src/libmfxsw_encode.cpp

```diff
@@ -54,77 +60,446 @@
 #include "mfx_vp9_encode_hw.h"
 #endif
 
-#if defined (MFX_ENABLE_HEVC_VIDEO_FEI_ENCODE)
 #if defined(MFX_VA)
 #include "mfx_h265_fei_encode_hw.h"
 #include <libmfx_core_interface.h>
 #endif
+
+
+struct CodecKey {
+    const mfxU32 codecId;
+    const bool   fei;
+
+    CodecKey(mfxU32 codecId, bool fei) : codecId(codecId), fei(fei) {}
+
+    friend bool operator<(CodecKey l, CodecKey r)
+    {
+        if (l.codecId == r.codecId)
+            return l.fei < r.fei;
+        return l.codecId < r.codecId;
+    }
+};
+
+struct Handlers {
+    struct Funcs {
+        std::function<VideoENCODE*(VideoCORE* core, mfxCoreInterface* coreInt, mfxStatus *mfxRes)> ctor;
+        std::function<mfxStatus(mfxSession s, mfxVideoParam *in, mfxVideoParam *out)> query;
+        std::function<mfxStatus(mfxSession s, mfxVideoParam *par, mfxFrameAllocRequest *request)> queryIOSurf;
+    };
+
+    Funcs primary;
+#if defined(MFX_ENABLE_SW_FALLBACK)
+    Funcs fallback;
 #endif
+};
 
-template<>
-VideoENCODE* _mfxSession::Create<VideoENCODE>(mfxVideoParam& par)
+typedef std::map<CodecKey, Handlers> CodecId2Handlers;
+
+enum {
+    MFX_CODEC_DUMMY_FOR_PLUGIN = 0
+};
+
+static const CodecId2Handlers codecId2Handlers =
 {
-    VideoENCODE* pENCODE = nullptr;
-    VideoCORE* core = m_pCORE.get();
-    mfxStatus mfxRes = MFX_ERR_MEMORY_ALLOC;
-    mfxU32 CodecId = par.mfx.CodecId;
-    // create a codec instance
-    switch (CodecId)
+#if defined(MFX_ENABLE_USER_ENCODE)
```

Comment:
leave a comment to warn everyone that this entry should be first

Reply:
I'm lost. Could you please clarify why it should be? To me, order is not important, I never relay on it.

Reply:
This is exactly logic I have. If `session->m_plgEnc.get()`, we use special `MFX_CODEC_DUMMY_FOR_PLUGIN`, and guarantee to have right handler (this handler is codec-independent). If not, other appropriate handlers are called. Why should we order something? We always know what to call.